zio-http/src/main/scala/zio/http/Flash.scala

```diff
@@ -0,0 +1,390 @@
+package zio.http
+
+import java.net.{URLDecoder, URLEncoder}
+import java.util.UUID
+
+import zio._
+
+import zio.schema.Schema
+import zio.schema.codec.JsonCodec
+
+import zio.http.template._
+
+/**
+ * `Flash` represents a flash value that one can retrieve from the flash scope.
+ *
+ * The flash scope consists of a serialized and url-encoded json object built
+ * with `zio-schema`.
+ */
+sealed trait Flash[+A] { self =>
+
+  def flatMap[B](f: A => Flash[B]): Flash[B] = Flash.FlatMap(self, f)
+
+  def map[B](f: A => B): Flash[B] = self.flatMap(a => Flash.succeed(f(a)))
+
+  def orElse[B >: A](that: => Flash[B]): Flash[B] = Flash.OrElse(self, that)
+
+  /**
+   * Operator alias for `orElse`.
+   */
+  def <>[B >: A](that: => Flash[B]): Flash[B] = self.orElse(that)
+
+  def zip[B](that: => Flash[B]): Flash[(A, B)] = self.zipWith(that)((a, b) => a -> b)
+
+  /**
+   * Operator alias for `zip`.
+   */
+  def <*>[B](that: => Flash[B]): Flash[(A, B)] = self.zip(that)
+
+  def zipWith[B, C](that: => Flash[B])(f: (A, B) => C): Flash[C] =
+    self.flatMap(a => that.map(b => f(a, b)))
+
+  def optional: Flash[Option[A]] = self.map(Option(_)) <> Flash.succeed(None)
+
+  def foldHtml[A1 >: A, B](f: Html => B, g: Html => B)(h: (B, B) => B)(implicit
+    ev: A1 =:= Flash.Message[Html, Html],
+  ): Flash[B] =
+    self.map(a => a.asInstanceOf[A1].fold(f, g)(h))
+
+  def toHtml[A1 >: A](implicit ev: A1 =:= String): Flash[Html] =
+    self.map(Html.fromString(_))
```

Comment:
Don't we need these methods to be `final`? Is there a need for them to be overridable?

Reply:
I can't hurt to make them final!